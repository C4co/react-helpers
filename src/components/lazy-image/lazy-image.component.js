import React, { useEffect, useRef } from "react"
import { isInViewport } from "../../dom"

/*
  name: LazyImage
  @loader -> string - image url to show at the first time
  @source -> string - image url to show when appear in view port
*/

export function LazyImage(props){
  const imageRef = useRef(null)

  function reload(){
    const item = imageRef.current

    function action(){
      if (isInViewport(item)) {
        item.src = props.source

        window.removeEventListener("scroll", action)
        window.removeEventListener("resize", action)
      }
    }

    window.addEventListener("scroll", action)
    window.addEventListener("resize", action)

    action()
  }

  useEffect(() => {
    reload()
  })

  return(
    <img
      {...props}
      ref={imageRef}
      src={props.loader}
      style={{
        ...props.style,
      }}
      alt={props.alt || ""}
    />
  )
}
