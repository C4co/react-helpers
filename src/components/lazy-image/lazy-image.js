import React, { useEffect, useRef } from "react"
import PropTypes from "prop-types"

export function LazyImage({ loader, style, alt, source, ...props }) {
  const imageRef = useRef(null)

  function isInViewport(elem) {
    const bounding = elem.getBoundingClientRect()

    return (
      bounding.top <= document.documentElement.clientHeight
    )
  }

  function reload() {
    const item = imageRef.current

    function action() {
      if (isInViewport(item)) {
        item.src = source

        window.removeEventListener("scroll", action)
        window.removeEventListener("resize", action)
      }
    }

    window.addEventListener("scroll", action)
    window.addEventListener("resize", action)

    action()
  }

  useEffect(() => {
    reload()
  })

  return (
    <img
      {...props}
      ref={imageRef}
      src={loader}
      style={{
        ...style
      }}
      alt={alt || ""}
    />
  )
}

LazyImage.propTypes = {
  /**  Image url */
  source: PropTypes.string,

  /** Pre-load image */
  loader: PropTypes.string,

  /** Image description */
  alt: PropTypes.string,

  /** CSS style */
  style: PropTypes.object
}

LazyImage.defaultProps = {
  source: "",
  loader: "",
  alt: "",
  style: {}
}
