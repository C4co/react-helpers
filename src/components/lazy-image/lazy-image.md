## Lazy Image

Just load a image when it appear in view port.

#### Usage
```jsx
  import { LazyImage } from "@cn/react-helpers"
```

#### Example:

```jsx
  <LazyImage
    style={{width: "100%"}}
    alt="image"
    loader={require("../images/loader.svg")}
    source={require("../images/image.jpg")}
  />
```

#### Properties:

| Name | Type | Default |
| --- | --- | --- |
| @loader | string(image path) | - |
| @source | string(image path) | - |
