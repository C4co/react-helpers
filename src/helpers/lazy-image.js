import React, { Component } from "react"

/*
  name: LazyImage
  @loader -> string - image url to show at the first time
  @source -> string - image url to show when appear in view port
*/
export class LazyImage extends Component {
  constructor(props) {
    super(props)
    this.myRef = React.createRef()
  }

  isInViewport(elem){
    var bounding = elem.getBoundingClientRect()
    const offset = 200

    return (
      bounding.top >= 0 &&
      bounding.left >= 0 &&
      bounding.bottom - offset <=
        (window.innerHeight || document.documentElement.clientHeight) &&
      bounding.right <=
        (window.innerWidth || document.documentElement.clientWidth)
    )
  }

  reload(){
    const item = this.myRef.current

    const action = () => {
      if (this.isInViewport(item)) {
        item.src = this.props.source
        item.style.opacity = "1"

        window.removeEventListener("scroll", action)
        window.removeEventListener("resize", action)
      }
    }

    window.addEventListener("scroll", action)
    window.addEventListener("resize", action)

    action()
  }

  componentDidMount() {
    this.reload()
  }

  render() {
    return (
      <img
        {...this.props}
        ref={this.myRef}
        src={this.props.loader}
        style={{
          opacity: "0.3",
          background: "red",
          transition: "all 0.3s cubic-bezier(0.215, 0.61, 0.355, 1) 0s",
          ...this.props.style,
        }}
        alt={this.props.alt || ""}
      />
    )
  }
}
