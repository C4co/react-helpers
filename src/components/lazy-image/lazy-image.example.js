import React from "react"
import {LazyImage} from "./lazy-image"
import styled from "styled-components"

const Wrapper = styled.section`
  margin: 0 auto;
  max-width: 500px;
  width: 100%;
`

function Example(){
  return(
  <Wrapper>
    <LazyImage
      style={{ width: "100%" }}
      alt="image"
      loader={require("./loader.svg")}
      source="https://cdn.pixabay.com/photo/2019/07/21/03/42/blue-4351901_960_720.jpg"
    />

    <LazyImage
      style={{ width: "100%" }}
      alt="image"
      loader={require("./loader.svg")}
      source="https://cdn.pixabay.com/photo/2013/12/17/20/10/air-bubbles-230014_960_720.jpg"
    />

    <LazyImage
      style={{ width: "100%" }}
      alt="image"
      loader={require("./loader.svg")}
      source="https://cdn.pixabay.com/photo/2015/03/26/09/47/sky-690293_960_720.jpg"
    />

    <LazyImage
      style={{ width: "100%" }}
      alt="image"
      loader={require("./loader.svg")}
      source="https://cdn.pixabay.com/photo/2016/03/04/19/36/beach-1236581_960_720.jpg"
    />

    <LazyImage
      style={{ width: "100%" }}
      alt="image"
      loader={require("./loader.svg")}
      source="https://cdn.pixabay.com/photo/2014/08/15/11/29/sea-418742_960_720.jpg"
    />

    <LazyImage
      style={{ width: "100%" }}
      alt="image"
      loader={require("./loader.svg")}
      source="https://cdn.pixabay.com/photo/2016/05/05/02/37/sunset-1373171_960_720.jpg"
    />
  </Wrapper>
  )
}

export default Example

