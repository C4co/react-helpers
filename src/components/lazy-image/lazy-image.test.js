import { mount } from "enzyme"
import { LazyImage } from "./lazy-image"
import React from "react"

const image = require("./image.png")
const loader = require("./loader.png")

describe("LazyImage component", () => {
  test("Check loader image", () => {
    const imgComponent = mount(
      <LazyImage
        style={{ width: "100%" }}
        alt='image'
        loader={loader}
        source={image}
      />
    )

    expect(imgComponent.find("img").prop("src")).toBe("loader.png")
    expect(imgComponent).toMatchSnapshot()
  })
})
