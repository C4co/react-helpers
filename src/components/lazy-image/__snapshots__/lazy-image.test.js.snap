// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LazyImage component Check loader image 1`] = `
<LazyImage
  alt="image"
  loader="loader.png"
  source="image.png"
  style={
    Object {
      "width": "100%",
    }
  }
>
  <img
    alt="image"
    src="loader.png"
    style={
      Object {
        "width": "100%",
      }
    }
  />
</LazyImage>
`;
